src/Simplex/Chat.hs

```diff
@@ -3425,13 +3427,46 @@ userProfileToSend user@User {profile = p} incognitoProfile ct =
       userPrefs = maybe (preferences' user) (const Nothing) incognitoProfile
    in (p' :: Profile) {preferences = Just . toChatPrefs $ mergePreferences (userPreferences <$> ct) userPrefs}
 
-createFeatureChangedItems :: (MsgDirectionI d, ChatMonad m) => User -> Contact -> Contact -> (Contact -> ChatDirection 'CTDirect d) -> (ChatFeature -> PrefEnabled -> Maybe Int -> CIContent d) -> m ()
-createFeatureChangedItems user Contact {mergedPreferences = cups} ct'@Contact {mergedPreferences = cups'} chatDir ciContent =
-  forM_ allChatFeatures $ \(ACF f) -> do
-    let state = featureState $ getContactUserPreference f cups
-        state' = featureState $ getContactUserPreference f cups'
-    when (state /= state') $
-      createInternalChatItem user (chatDir ct') (uncurry (ciContent $ chatFeature f) state') Nothing
+createRcvFeatureItems :: forall m. ChatMonad m => User -> Contact -> Contact -> m ()
+createRcvFeatureItems user ct ct' =
+  createFeatureItems user ct ct' CDDirectRcv CIRcvChatFeature CIRcvFeatureOffer contactPreference
+
+createSndFeatureItems :: forall m. ChatMonad m => User -> Contact -> Contact -> m ()
+createSndFeatureItems user ct ct' =
+  createFeatureItems user ct ct' CDDirectSnd CISndChatFeature CISndFeatureOffer getPref
+  where
+    getPref = (preference :: ContactUserPref (FeaturePreference f) -> FeaturePreference f) . userPreference
+
+createFeatureItems ::
+  forall d m.
+  (MsgDirectionI d, ChatMonad m) =>
+  User ->
+  Contact ->
+  Contact ->
+  (Contact -> ChatDirection 'CTDirect d) ->
+  (ChatFeature -> PrefEnabled -> Maybe Int -> CIContent d) ->
+  (ChatFeature -> Maybe Int -> CIContent d) ->
+  (forall f. ContactUserPreference (FeaturePreference f) -> FeaturePreference f) ->
+  m ()
+createFeatureItems user Contact {mergedPreferences = cups} ct'@Contact {mergedPreferences = cups'} chatDir ciFeature ciOffer getPref =
+  forM_ allChatFeatures $ \(ACF f) -> createItem f
+  where
+    createItem :: forall f. FeatureI f => SChatFeature f -> m ()
+    createItem f
+      | state /= state' = create $ uncurry (ciFeature f') state'
+      | offer = create $ ciOffer f' param'
+      | otherwise = pure ()
+      where
+        create content = createInternalChatItem user (chatDir ct') content Nothing
+        f' = chatFeature f
+        state = featureState cup
+        state' = featureState cup'
+        offer = (allow' == FAYes || allow' == FAAlways) && (allow == FANo || param /= param')
```

Comment:
why does this affect only disappearing messages in tests? shouldn't it create offers for other features as well?

Reply:
I think I understand, it's because other features have items created on state /= state' condition, so they won't have interactive offer.. not sure it's desired behaviour

Reply:
one-way state changes can be interactive too